### PR TITLE
Custom deserializer - allow to deserialize dates or any other type yourself.

### PR DESCRIPTION
**Pull request by rflament**
Custom deserializer - allow to deserialize dates or any other type yourself with any external library you want.

This is another way to do what is in #840 without the need to add an new dependency (moment.js) in the codebase.

Instead of modifying node-soap to use moment.js to deserialize dates, it adds the possibility to give node-soap a custom function to use for a type.

This way the code using node-soap can use moment.js to deserialize dates, or just pass dates as string.
It can also be used for other types, booleans, integers, etc.

The context can be used to differentiate xml element by their names.

### CODE REVIEW

**Comment by rflament** [comment]
@jsdevel are you ok with this one ?

File: test/request-response-samples/RetrieveCustomerDetails__returning_a_date_should_parse_the_date/soap.wsdl [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                <xsd:complexType>
+                    <xsd:sequence>
+                        <xsd:element maxOccurs="1" minOccurs="0" name="accountNumber" nillable="false"
+                                     type="xsd:string"/>
+                        <xsd:element minOccurs="0" name="lastUpdatedDate" type="xsd:date"/>
+                    </xsd:sequence>
+                </xsd:complexType>
+            </xsd:element>
+
+
```
Comment (jsdevel) [inline test/request-response-samples/RetrieveCustomerDetails__returning_a_date_should_parse_the_date/soap.wsdl]: can you remove this excess space?

File: test/request-response-samples/RetrieveCustomerDetails__returning_a_date_should_parse_the_date/wsdl_options.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+exports.customDeserializer = {
+
+  date: function (text, context) {
+    return text;
+  }
+};
+
+exports.ignoreBaseNameSpaces = false;
+
```
Comment (jsdevel) [inline test/request-response-samples/RetrieveCustomerDetails__returning_a_date_should_parse_the_date/wsdl_options.js]: can you remove these line endings?

File: Readme.md [hunk 3]
```diff
@@ -445,6 +445,8 @@ Soap body contents. Useful if you don't want to log /store Soap headers.
 the SOAP response body as well as the entire `IncomingMessage` response object.
 This is emitted for all responses (both success and errors).
 
+
```
Comment (jsdevel) [inline Readme.md]: can you remove this?

File: lib/wsdl.js [hunk 4]
```diff
@@ -1422,18 +1423,23 @@ WSDL.prototype.xmlToObject = function(xml, callback) {
...
-      value = parseInt(text, 10);
-    } else if (name === 'bool' || name === 'boolean') {
-      value = text.toLowerCase() === 'true' || text === '1';
-    } else if (name === 'dateTime' || name === 'date') {
-      value = new Date(text);
-    } else {
-      // handle string or other types
-      if (typeof top.object !== 'string') {
-        value = text;
+    if (self.options && self.options.customDeserializer && self.options.customDeserializer[ name ]) {
```
Comment (jsdevel) [inline lib/wsdl.js]: no space between brackets for property access E.G. `self.options.customDeserializer[name]`

**Comment by rflament** [comment]
@jsdevel all done !

**Comment by jsdevel** [comment]
Seems fine to me.  Thoughts @herom ?

**Comment by herom** [comment]
I'm all 👍 for this 😃

**Pull request merged**